src/Oracle/Protocol/OracleHttpProtocol.cs

```diff
@@ -0,0 +1,92 @@
+using System;
+using System.Data;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Threading.Tasks;
+using JArray = Newtonsoft.Json.Linq.JArray;
+using JObject = Newtonsoft.Json.Linq.JObject;
+
+namespace Neo.Plugins
+{
+    public class OracleHttpProtocol : IOracleProtocol
+    {
+        public int Timeout { get; set; } = 5000;
+        public bool AllowPrivateHost { get; set; } = false;
+        public readonly string[] AllowedFormats = new string[] { "application/json" };
+
+        public byte[] Request(ulong requestId, string url, string filter)
+        {
+            Utility.Log(nameof(OracleHttpProtocol), LogLevel.Debug, $"Request: {url}");
+
+            if (!Uri.TryCreate(url, UriKind.Absolute, out var uri)) throw new InvalidOperationException("UrlError");
+            if (!AllowPrivateHost && IsInternal(Dns.GetHostEntry(uri.Host))) throw new InvalidOperationException("Access to private host is not allowed");
+
+            using var handler = new HttpClientHandler();
+            using var client = new HttpClient(handler);
+            client.DefaultRequestHeaders.Add("Accept", string.Join(",", AllowedFormats));
+
+            Task<HttpResponseMessage> result = client.GetAsync(uri);
+            Stopwatch sw = new Stopwatch();
+            sw.Start();
+            if (!result.Wait(Timeout)) throw new TimeoutException("Timeout");
+            if (result.Result.StatusCode == HttpStatusCode.NotFound) throw new FileNotFoundException($"{url} is not found");
+            if (!result.Result.IsSuccessStatusCode) throw new InvalidOperationException("Response error");
+            if (!AllowedFormats.Contains(result.Result.Content.Headers.ContentType.MediaType)) throw new InvalidOperationException("ContentType it's not allowed");
+            sw.Stop();
+            var taskRet = result.Result.Content.ReadAsStringAsync();
+            if (Timeout <= sw.ElapsedMilliseconds || !taskRet.Wait(Timeout - (int)sw.ElapsedMilliseconds)) throw new TimeoutException("Timeout");
+            var data = Filter(taskRet.Result, filter);
+            return Utility.StrictUTF8.GetBytes(data);
+        }
+
+        private string Filter(string input, string filterArgs)
+        {
+            if (string.IsNullOrEmpty(filterArgs))
+                return input;
+
+            JObject beforeObject = JObject.Parse(input);
+            JArray afterObjects = new JArray(beforeObject.SelectTokens(filterArgs, true).ToArray());
```

Comment:
I don't think you need the ToArray call here: https://stackoverflow.com/a/44839613

Reply:
Done